rdkPlugins/Storage/source/BindLoopMountDetails.cpp

```diff
@@ -0,0 +1,459 @@
+* limitations under the License.
+*/
+
+#include "BindLoopMountDetails.h"
+#include "StorageHelper.h"
+#include "DobbyRdkPluginUtils.h"
+#include "RefCountFile.h"
+#include "RefCountFileLock.h"
+
+#include <fstream>
+#include <stdio.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sstream>
+#include <fstream>
+#include <dirent.h>
+
+
+#if defined(__linux__)
+#include <linux/loop.h>
+#endif
+
+
+// Loop mount details class
+BindLoopMountDetails::BindLoopMountDetails(const std::string& rootfsPath,
+                                   const MountProperties& mount,
+                                   const uid_t& userId,
+                                   const gid_t& groupId,
+                                   const std::shared_ptr<DobbyRdkPluginUtils> &utils)
+    : mMount(mount),
+    mUserId(userId),
+    mGroupId(groupId),
+    mUtils(utils)
+{
+    AI_LOG_FN_ENTRY();
+
+    mMountPointInsideContainer = rootfsPath + mount.destination;
+    mTempMountPointOutsideContainer = mMountPointInsideContainer + ".temp";
+
+    AI_LOG_FN_EXIT();
+}
+
+BindLoopMountDetails::~BindLoopMountDetails()
+{
+    AI_LOG_FN_ENTRY();
+
+    AI_LOG_FN_EXIT();
+}
+
+// -----------------------------------------------------------------------------
+/**
+ *  @brief Opens the data.img file and mounts it to temp location inside container
+ *
+ *  @return true on success, false on failure.
+ */
+bool BindLoopMountDetails::onPreCreate()
+{
+    AI_LOG_FN_ENTRY();
+    bool success = false;
+
+    // step 1 - Create image file if not yet there, this also runs checks on
+    // the integrity of the image, those should be enabled if imgManagement
+    // is also enabled (the default is true)
+    if (mMount.imgManagement)
+    {
+        if (!StorageHelper::createFileIfNeeded(mMount.fsImagePath,
+                                               mMount.imgSize,
+                                               mUserId,
+                                               mMount.fsImageType))
+        {
+            // logging already provided by createFileIfNeeded
+            return false;
+        }
+    }
+    else if (access(mMount.fsImagePath.c_str(), F_OK) != 0)
+    {
+        AI_LOG_ERROR("failed to find source image '%s' for storage plugin",
+                     mMount.fsImagePath.c_str());
+        return false;
+    }
+
+    // Open reference count file and lock it
+    auto refCountFile = getRefCountFile();
+    if (!refCountFile)
+    {
+        return false;
+    }
+    RefCountFileLock lock(refCountFile);
+
+    // step 2 - if loop device doesn't exist, try and open the source file and
+    // attach to a space loop device 
+    std::string loopDevice = StorageHelper::getLoopDevice(mMount.fsImagePath);
+    int loopDevFd(0);
+
+    if (loopDevice.empty())
+    {
+        loopDevFd = StorageHelper::attachLoopDevice(mMount.fsImagePath, &loopDevice);
```

Comment:
Should we log an error if the reference count is greater than one and loop device doesn't exist.

Reply:
Possible. If the app crashed this could happen. Therefore, we need a way to reset the reference count.

Reply:
It shouldn't happen normally, but I don't we should abandon launching the app.